File: 2024-11-November-GFG-POTD/November 06 - Root to Leaf Paths sum/Solution.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class Solution:
    def treePathsSum(self, root):
        if root is None:
            return 0

        total_sum = 0
        stack = [(root, root.data)]

        while stack:
            node, current_sum = stack.pop()

            if node.left is None and node.right is None:
                total_sum += current_sum

            if node.right:
                stack.append((node.right, current_sum * 10 + node.right.data))
            if node.left:
                stack.append((node.left, current_sum * 10 + node.left.data))

        return total_sum

def build_tree(s):
    if not s or s[0] == 'N':
        return None

    values = s.split()
    root = Node(int(values[0]))
    queue = [root]
    i = 1

    while queue and i < len(values):
        current = queue.pop(0)

        if values[i] != 'N':
            current.left = Node(int(values[i]))
            queue.append(current.left)
        i += 1

        if i >= len(values):
            break

        if values[i] != 'N':
            current.right = Node(int(values[i]))
            queue.append(current.right)
        i += 1

    return root

if __name__ == "__main__":
    t = int(input().strip())
    for _ in range(t):
        s = input().strip()
        root = build_tree(s)
        ob = Solution()
        print(ob.treePathsSum(root))

File: 2024-11-November-GFG-POTD/November 06 - Root to Leaf Paths sum/Solution.cpp
#include <iostream>
#include <stack>
#include <queue>
#include <string>
#include <sstream>
using namespace std;

struct Node {
    int data;
    Node* left;
    Node* right;

    Node(int d) {
        data = d;
        left = nullptr;
        right = nullptr;
    }
};

class Solution {
public:
    int treePathsSum(Node* root) {
        if (root == nullptr) return 0;

        int totalSum = 0;
        stack<pair<Node*, int>> stack;
        stack.push({root, root->data});

        while (!stack.empty()) {
            auto current = stack.top();
            stack.pop();
            Node* node = current.first;
            int currentSum = current.second;

            if (node->left == nullptr && node->right == nullptr) {
                totalSum += currentSum;
            }

            if (node->right != nullptr) {
                stack.push({node->right, currentSum * 10 + node->right->data});
            }
            if (node->left != nullptr) {
                stack.push({node->left, currentSum * 10 + node->left->data});
            }
        }

        return totalSum;
    }
};

Node* buildTree(const string& str) {
    if (str.length() == 0 || str[0] == 'N') return nullptr;

    istringstream iss(str);
    string val;
    iss >> val;
    Node* root = new Node(stoi(val));
    queue<Node*> queue;
    queue.push(root);

    while (iss >> val) {
        Node* currNode = queue.front();
        queue.pop();

        if (val != "N") {
            currNode->left = new Node(stoi(val));
            queue.push(currNode->left);
        }

        if (!(iss >> val)) break;
        if (val != "N") {
            currNode->right = new Node(stoi(val));
            queue.push(currNode->right);
        }
    }

    return root;
}

int main() {
    int t;
    cin >> t;
    cin.ignore();
    while (t--) {
        string s;
        getline(cin, s);
        Node* root = buildTree(s);
        Solution ob;
        cout << ob.treePathsSum(root) << endl;
    }
    return 0;
}
